### DISCUSSION TITLE
sort column in ui.table but consider blank value in special way?

### DISCUSSION THREAD

**Discussion by retsyo** [zauberzeug/nicegui#2238] [Q&A] [answered]
### Question

There are 2 cases. Why did I pose 2 cases? I don't know which cases is more common/natural during sorting in various applications, sorry.

# 1. case 1, if a cell has an omitted value, how to let it go behind 0 during sorting? 

For example, in the following code, if the `Score` value is sorted from big to small, I expect the numerical values have a sequence like
```
100.00 > 88.00 > 66.50 > 0.00 > blank = blank 
```
or in other words, `blank` is treated as a negative value during sorting.

but the following code can only get
```
blank = blank > 100.00 > 88.00 > 66.50 > 0.00
```

BYW, since the scores are kept two decimal places, they are string in the following code

```python
columns = [
    {
        'name': 'name',
        'label': 'Name',
        'field': 'name',
    },
    {
        'name': 'score',
        'label': 'Score',
        'field': 'score',
        'sortable': True,
        ':sort': '(a, b, rowA, rowB) => Number(b) - parseFloat(a)',
    },
]
rows = [
    {'name': 'Bob', 'score': '88.00'},
    {'name': 'Alice', 'score': '100.00'},
    {'name': 'Bobby', 'score': '66.50'},
    {'name': 'Mike', 'score': '0.00'},
    {'name': 'Carl'},
    {'name': 'John'},
]
ui.table(columns=columns, rows=rows, row_key='name')
```

thanks

**Comment by retsyo** [comment]
# 2. case 2, if a cell has an omitted value, do not add it into sorting processing?
I mean, if the `Score` value is sorted from big to small, the result should be
```
100.00 > 88.00 > 66.50 > 0.00  blank = blank 
```
while if the `Score` value is sorted from small to big, the result should be
```
0.00 < 66.50 < 88.00 < 100.00  blank = blank
```

**Comment by retsyo** [comment]
> # 1. case 1, if a cell has an omitted value, how to let it go behind 0 during sorting?

I get a solution for this case
```python
':sort': '(a, b) => (Number(a)||(-1)) - (parseFloat(b)||(-1))'
```

**Comment by falkoschindler** [comment, marked as answer]
Case 2 is tricky, because you need to consider the current sort direction, which is not available in the `:sort` function. But you can define a custom sort method like this:
```py
columns = [
    {'name': 'name', 'label': 'Name', 'field': 'name'},
    {'name': 'score', 'label': 'Score', 'field': 'score', 'sortable': True},
]
rows = [
    {'name': 'Bob', 'score': '88.00'},
    {'name': 'Alice', 'score': '100.00'},
    {'name': 'Bobby', 'score': '66.50'},
    {'name': 'Mike', 'score': '0.00'},
    {'name': 'Carl'},
    {'name': 'John'},
]
table = ui.table(columns=columns, rows=rows, row_key='name')
table._props[':sort-method'] = '''(rows, sortBy, descending) => {
    const data = [...rows];
    if (sortBy) {
        data.sort((a, b) => {
            if (a[sortBy] === undefined) return 1;
            if (b[sortBy] === undefined) return -1;
            const result = parseFloat(a[sortBy]) - parseFloat(b[sortBy]);
            return descending ? -result : result;
        });
    }
    return data;
}
'''
```
